File: amp-ai/python/runtime/agent.py
from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import anthropic
import yaml

ROOT = Path(__file__).resolve().parents[2]

MODEL_MAP = {
    "claude-opus-4": "claude-opus-4-7",
    "claude-opus-4-7": "claude-opus-4-7",
    "claude-opus-4-6": "claude-opus-4-6",
    "claude-sonnet-4": "claude-sonnet-4-6",
    "claude-sonnet-4-6": "claude-sonnet-4-6",
    "claude-haiku-4-5": "claude-haiku-4-5",
}


def resolve_model(name: str | None) -> str:
    if not name:
        return os.environ.get("AMP_DEFAULT_MODEL", "claude-opus-4-7")
    return MODEL_MAP.get(name, name)


@dataclass
class LoadedAgent:
    slug: str
    config: dict[str, Any]
    system_prompt: str


_client: anthropic.Anthropic | None = None


def _get_client() -> anthropic.Anthropic:
    global _client
    if _client is None:
        if not os.environ.get("ANTHROPIC_API_KEY"):
            raise RuntimeError(
                "ANTHROPIC_API_KEY is not set. Copy .env.example to .env and fill it in."
            )
        _client = anthropic.Anthropic()
    return _client


def load_agent(slug: str) -> LoadedAgent:
    agent_dir = ROOT / "agents" / slug
    system_prompt = (agent_dir / "system.md").read_text(encoding="utf-8")
    config = yaml.safe_load((agent_dir / "config.yaml").read_text(encoding="utf-8"))
    return LoadedAgent(slug=slug, config=config, system_prompt=system_prompt)


def list_agents() -> list[str]:
    return sorted(p.name for p in (ROOT / "agents").iterdir() if p.is_dir())


def _build_user_message(user_input: str, results: dict[str, str]) -> str:
    if not results:
        return user_input
    upstream = "\n\n".join(
        f"### Upstream agent: {name}\n{output}" for name, output in results.items()
    )
    return f"# User request\n{user_input}\n\n# Upstream agent outputs\n{upstream}"


def run_agent(agent: LoadedAgent, user_input: str, results: dict[str, str]) -> str:
    if agent.config.get("engine") or agent.slug == "image-generator":
        from .tools import generate_render

        prompt = _build_user_message(user_input, results)
        result = generate_render(
            prompt,
            engine=agent.config.get("engine"),
            resolution=agent.config.get("resolution"),
            quality=agent.config.get("quality"),
        )
        import json

        return json.dumps(result, indent=2)

    model = resolve_model(agent.config.get("model"))
    client = _get_client()

    with client.messages.stream(
        model=model,
        max_tokens=16000,
        thinking={"type": "adaptive"},
        system=[
            {
                "type": "text",
                "text": agent.system_prompt,
                "cache_control": {"type": "ephemeral"},
            }
        ],
        messages=[{"role": "user", "content": _build_user_message(user_input, results)}],
    ) as stream:
        message = stream.get_final_message()

    return "\n".join(block.text for block in message.content if block.type == "text")

File: amp-ai/runtime/agent.ts
import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import Anthropic from '@anthropic-ai/sdk';
import YAML from 'yaml';
import type { AgentConfig, LoadedAgent, RunContext } from './types.ts';
import { resolveModel } from './model-map.ts';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');

let client: Anthropic | null = null;
function getClient(): Anthropic {
  if (!client) {
    if (!process.env.ANTHROPIC_API_KEY) {
      throw new Error('ANTHROPIC_API_KEY is not set. Copy .env.example to .env and fill it in.');
    }
    client = new Anthropic();
  }
  return client;
}

export async function loadAgent(slug: string): Promise<LoadedAgent> {
  const dir = path.join(ROOT, 'agents', slug);
  const [systemPrompt, configRaw] = await Promise.all([
    fs.readFile(path.join(dir, 'system.md'), 'utf-8'),
    fs.readFile(path.join(dir, 'config.yaml'), 'utf-8'),
  ]);
  const config = YAML.parse(configRaw) as AgentConfig;
  return { slug, config, systemPrompt };
}

export async function listAgents(): Promise<string[]> {
  const dir = path.join(ROOT, 'agents');
  const entries = await fs.readdir(dir, { withFileTypes: true });
  return entries.filter((e) => e.isDirectory()).map((e) => e.name);
}

function buildUserMessage(ctx: RunContext): string {
  if (Object.keys(ctx.results).length === 0) return ctx.userInput;
  const upstream = Object.entries(ctx.results)
    .map(([name, out]) => `### Upstream agent: ${name}\n${out}`)
    .join('\n\n');
  return `# User request\n${ctx.userInput}\n\n# Upstream agent outputs\n${upstream}`;
}

export async function runAgent(agent: LoadedAgent, ctx: RunContext): Promise<string> {
  if (agent.config.engine || agent.slug === 'image-generator') {
    const { generateRender } = await import('../tools/render-engine.ts');
    const promptText = buildUserMessage(ctx);
    const result = await generateRender(promptText, {
      engine: agent.config.engine,
      resolution: agent.config.resolution,
      quality: agent.config.quality,
    });
    return JSON.stringify(result, null, 2);
  }

  const model = resolveModel(agent.config.model);
  const stream = await getClient().messages.stream({
    model,
    max_tokens: 16000,
    thinking: { type: 'adaptive' },
    system: [
      {
        type: 'text',
        text: agent.systemPrompt,
        cache_control: { type: 'ephemeral' },
      },
    ],
    messages: [{ role: 'user', content: buildUserMessage(ctx) }],
  });

  const message = await stream.finalMessage();
  return message.content
    .filter((b): b is Anthropic.TextBlock => b.type === 'text')
    .map((b) => b.text)
    .join('\n');
}
